File: bitmarket-common/src/main/java/com/bitwise/bitmarket/common/currency/BtcAmount.java
package com.bitwise.bitmarket.common.currency;

import java.math.BigDecimal;

public class BtcAmount {

    private final BigDecimal amount;

    public BtcAmount(BigDecimal amount) {
        super();
        this.amount = amount;
    }

    public BigDecimal getAmount() {
        return this.amount;
    }

    @Override
    public String toString() {
        return String.format("%s BTC", this.amount.toPlainString());
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + (this.amount == null ? 0 : this.amount.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (this.getClass() != obj.getClass()) {
            return false;
        }
        final BtcAmount other = (BtcAmount) obj;
        if (this.amount == null) {
            if (other.amount != null) {
                return false;
            }
        } else if (!this.amount.equals(other.amount)) {
            return false;
        }
        return true;
    }
}

File: bitmarket-common/src/main/scala/com/bitwise/bitmarket/common/currency/BtcAmount.scala
package com.bitwise.bitmarket.common.currency

case class BtcAmount(amount: BigDecimal) extends Ordered[BtcAmount] {

  /** Alternative constructor for Java code */
  def this(amount: java.math.BigDecimal) = this(new BigDecimal(amount))

  def min(other: BtcAmount): BtcAmount = if (this <= other) this else other
  def +(that: BtcAmount) = BtcAmount(amount + that.amount)
  def -(that: BtcAmount) = BtcAmount(amount - that.amount)
  def unary_- = BtcAmount(-amount)

  override def compare(that: BtcAmount): Int = this.amount.compare(that.amount)

  override def toString = amount.underlying().toPlainString + " BTC"
}

object BtcAmount extends Ordering[BtcAmount] {
  override def compare(x: BtcAmount, y: BtcAmount): Int = x.amount.compare(y.amount)
}
